lib/internal/process/per_thread.js

```diff
@@ -296,88 +373,562 @@ function buildAllowedFlags() {
     }
   }
 
-  const trimLeadingDashes =
-    (flag) => StringPrototypeReplace(flag, leadingDashesRegex, '');
-
-  // Save these for comparison against flags provided to
-  // process.allowedNodeEnvironmentFlags.has() which lack leading dashes.
-  const nodeFlags = ArrayPrototypeMap(allowedNodeEnvironmentFlags,
-                                      trimLeadingDashes);
-
-  class NodeEnvironmentFlagsSet extends Set {
-    constructor(array) {
-      super();
-      this[kInternal] = { array };
-    }
-
-    add() {
-      // No-op, `Set` API compatible
-      return this;
-    }
-
-    delete() {
-      // No-op, `Set` API compatible
-      return false;
-    }
-
-    clear() {
-      // No-op, `Set` API compatible
-    }
-
-    has(key) {
-      // This will return `true` based on various possible
-      // permutations of a flag, including present/missing leading
-      // dash(es) and/or underscores-for-dashes.
-      // Strips any values after `=`, inclusive.
-      // TODO(addaleax): It might be more flexible to run the option parser
-      // on a dummy option set and see whether it rejects the argument or
-      // not.
-      if (typeof key === 'string') {
-        key = StringPrototypeReplace(key, replaceUnderscoresRegex, '-');
-        if (RegExpPrototypeTest(leadingDashesRegex, key)) {
-          key = StringPrototypeReplace(key, trailingValuesRegex, '');
-          return ArrayPrototypeIncludes(this[kInternal].array, key);
-        }
-        return ArrayPrototypeIncludes(nodeFlags, key);
-      }
-      return false;
-    }
-
-    entries() {
-      this[kInternal].set ??=
-        new Set(new SafeArrayIterator(this[kInternal].array));
-      return SetPrototypeEntries(this[kInternal].set);
-    }
-
-    forEach(callback, thisArg = undefined) {
-      ArrayPrototypeForEach(
-        this[kInternal].array,
-        (v) => ReflectApply(callback, thisArg, [v, v, this])
-      );
-    }
-
-    get size() {
-      return this[kInternal].array.length;
-    }
-
-    values() {
-      this[kInternal].set ??=
-        new Set(new SafeArrayIterator(this[kInternal].array));
-      return SetPrototypeValues(this[kInternal].set);
-    }
-  }
-  NodeEnvironmentFlagsSet.prototype.keys =
-  NodeEnvironmentFlagsSet.prototype[SymbolIterator] =
-    NodeEnvironmentFlagsSet.prototype.values;
-
-  ObjectFreeze(NodeEnvironmentFlagsSet.prototype.constructor);
-  ObjectFreeze(NodeEnvironmentFlagsSet.prototype);
-
-  return ObjectFreeze(new NodeEnvironmentFlagsSet(
+  return ObjectFreeze(new FlagsSet(
     allowedNodeEnvironmentFlags
   ));
 }
 
+// This builds the initial process.availableV8Flags.
+function buildAvailableV8Flags() {
+
+  const v8Flags = [
+    '--abort-on-contradictory-flags',
```

Comment:
This doesn't look very reliable to me - V8 flags come and go and there are no guarantees about them. If possible we should do this programmatically, e.g. having a script that put them into a json file during every update and loading the json during build time, or just request v8 to expose a list of available flags through embedder API (that might be difficult to get accepted though).

Reply:
Yup, since the flags can come or go in any release, I added a test to check these flags against the output of `node --v8-options. Unfortunately, as if there wasn't already enough reason to be skeptical about a hardcoded list of flags, the test fails because some flags aren't available on some platforms. And that's not even getting into problems when Node.js is built with different configuration flags. So this won't work at all without a bunch of extra logic. So yeah, the hardcoded array is a total nonstarter which is not surprising.